### PR TITLE
Chapter1/homework/9

### PR DESCRIPTION
**Pull request by nevin75**

### CODE REVIEW

File: Src/BootCamp.Chapter/Conversions.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Conversions.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            int i;
+            int[] a = new int[40];
+            Console.Write("Enter the number to convert: ");
+            if (!int.TryParse(Console.ReadLine(), out int decimalNumber))
+            {
+                Console.WriteLine("That wasn't a decimal number. Please try again");
+                ConvertDecimalToBinary();
+            }
+            for (i = 0; decimalNumber > 0; i++)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [This loop's stop condition tests 'decimalNumber' but the incrementer updates 'i'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            for (i = i - 1; i >= 0; i--)
+            {
+                Console.Write(a[i]);
+            }
+        }
+
+        private static void ConvertBinaryToDecimal()
+        {
+            int decimalValue = 0;
+            int[] a = new int[10];
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove the unused local variable 'a'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/LetterConverter.cs  5
- Src/BootCamp.Chapter/Conversions.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

File: Src/BootCamp.Chapter/LetterConverter.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            return false;
+
+        }
+        private static char ConvertKeyToArrow (char keyStroke)
+        {
+            if (char.ToUpper(keyStroke) == 'W')
+                return '\u2191';
+            if (char.ToUpper(keyStroke) == 'S')
+                return '\u2193';
+            if (char.ToUpper(keyStroke) == 'A')
```
Comment (Almantask) [inline Src/BootCamp.Chapter/LetterConverter.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/LetterConverter.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            {
+                if (keyStroke == letter)
+                    return true;
+            }
+            return false;
+
+        }
+        private static char ConvertKeyToArrow (char keyStroke)
+        {
+            if (char.ToUpper(keyStroke) == 'W')
```
Comment (Almantask) [inline Src/BootCamp.Chapter/LetterConverter.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/LetterConverter.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/LetterConverter.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+=======
+                Console.Write(a[i]);
+>>>>>>> 3ccdb32e75cfac3de3437700b11d80795480c7f2
+            }
+        }
+
+        private static void ConvertBinaryToDecimal()
+        {
+            int decimalValue = 0;
+            int[] a = new int[10];
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove the unused local variable 'a'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/LetterConverter.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            Console.WriteLine("Please enter W, A, S or D to practice your First Person Shooter moves.");
+            return Console.ReadKey(true).KeyChar;
+        }
+
+        private static bool IsValidKey (char keyStroke)
+        {
+            char[] expectedKeys = { 'w', 'W', 'A', 'a', 's', 'S', 'd', 'D' }; 
+            foreach (char letter in expectedKeys)
+            {
+                if (keyStroke == letter)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/LetterConverter.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/LetterConverter.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        private static char ConvertKeyToArrow (char keyStroke)
+        {
+            if (char.ToUpper(keyStroke) == 'W')
+                return '\u2191';
+            if (char.ToUpper(keyStroke) == 'S')
+                return '\u2193';
+            if (char.ToUpper(keyStroke) == 'A')
+                return '\u2190';
+            if (char.ToUpper(keyStroke) == 'D')
```
Comment (Almantask) [inline Src/BootCamp.Chapter/LetterConverter.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/Program.cs [hunk 11]
```diff
@@ -1,12 +1,32 @@
...
+﻿
 namespace BootCamp.Chapter
 {
     class Program
     {
         static void Main(string[] args)
         {
-            Console.WriteLine("Hello World!");
+<<<<<<< HEAD
+            //Conversions.Demo();
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Remove this commented out code.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/LetterConverter.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+                    return true;
+            }
+            return false;
+
+        }
+        private static char ConvertKeyToArrow (char keyStroke)
+        {
+            if (char.ToUpper(keyStroke) == 'W')
+                return '\u2191';
+            if (char.ToUpper(keyStroke) == 'S')
```
Comment (Almantask) [inline Src/BootCamp.Chapter/LetterConverter.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/LetterConverter.cs  5
- Src/BootCamp.Chapter/Conversions.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

File: Src/BootCamp.Chapter/Conversions.cs (hunk 8 above)
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove the unused local variable 'a'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/LetterConverter.cs  5
- Src/BootCamp.Chapter/Conversions.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+                Console.WriteLine("That wasn't a decimal number. Please try again");
+                ConvertDecimalToBinary();
+            }
+            i = 0;
+            while (decimalNumber > 0)
+            {
+                binaryHolder[i] = decimalNumber % 2;
+                decimalNumber = decimalNumber / 2;
+                i++;
+                for (i = 0; decimalNumber > 0; i++)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [This loop's stop condition tests 'decimalNumber' but the incrementer updates 'i'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+                    Console.Write(binaryHolder[i]);
+                    Console.Write(a[i]);
+                }
+            }
+        }
+
+            private static void ConvertBinaryToDecimal()
+            {
+                int decimalValue = 0;
+                int[] a = new int[10];
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove the unused local variable 'a'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/LetterConverter.cs  5
- Src/BootCamp.Chapter/Conversions.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            Console.WriteLine();
+            Console.WriteLine("Convert Binary To Decimal");
+            ConvertBinaryToDecimal();
+        }
+
+        private static void ConvertDecimalToBinary()
+        {
+            int i;
+            int[] binaryHolder = new int[40];
+            int[] a = new int[40];
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove the unused local variable 'a'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+            for (i = i - 1; i >= 0; i--)
+            {
+                Console.Write(binaryHolder[i]);
+            }
+        }
+
+            private static void ConvertBinaryToDecimal()
+            {
+                int decimalValue = 0;
+                int[] a = new int[10];
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Codacy Issue found: [Remove the unused local variable 'a'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+                Console.Write("Enter the binary number to convert: ");
+                string binary = Console.ReadLine();
+                if (!Regex.IsMatch(binary, @"^[0-1]+$"))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Well done for using regex :)

File: Src/BootCamp.Chapter/Conversions.cs [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+            ConvertDecimalToBinary();
+            Console.WriteLine();
+            Console.WriteLine("Convert Binary To Decimal");
+            ConvertBinaryToDecimal();
+        }
+
+        private static void ConvertDecimalToBinary()
+        {
+            int i;
+            int[] binaryHolder = new int[40];
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: Why 40? It's a magic number, please explain why 40, preferably use named constant.

File: Src/BootCamp.Chapter/Conversions.cs [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        public static void Demo()
+        {
+            Console.WriteLine("Convert Decimal To Binary");
+            ConvertDecimalToBinary();
+            Console.WriteLine();
+            Console.WriteLine("Convert Binary To Decimal");
+            ConvertBinaryToDecimal();
+        }
+
+        private static void ConvertDecimalToBinary()
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Conversions.cs]: This function does more than 1 thing.
Split reading input and converting part.
You can still use both in one function, it should just be split in at least 2 parts.
And it is more natural that was, we convert something, so it makes sense that we pass input to convert it.

File: Src/BootCamp.Chapter/Conversions.cs (hunk 18 above)
Comment (nevin75) [reply Src/BootCamp.Chapter/Conversions.cs]: I'm creating an array to hold the "string" of bits that represent the decimal number. I don't know how to do this without pre-naming the length of the array. 40 is an arbitrary number, though. It seems like there must be a better, more dynamic way. I don't know what it is.

I think "magic number" is whenever I'm defining a static value in my algorithm, I should define it as a constant at the top of the code so it is more easily found and edited. Thanks for that. I'll make changes wherever I have magic numbers.

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/LetterConverter.cs  5
- Src/BootCamp.Chapter/Conversions.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

File: Src/BootCamp.Chapter/Program.cs [hunk 20]
```diff
@@ -1,12 +1,12 @@
...
-
+﻿
 namespace BootCamp.Chapter
 {
     class Program
     {
         static void Main(string[] args)
         {
-            Console.WriteLine("Hello World!");
+            //Conversions.Demo();
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Remove this commented out code.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/ArrowMovement.cs  4
- Src/BootCamp.Chapter/BinaryConverter.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/ArrowMovement.cs  4
- Src/BootCamp.Chapter/BinaryConverter.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723&bid=16529421)

File: Src/BootCamp.Chapter/Program.cs [hunk 21]
```diff
@@ -1,12 +1,10 @@
-﻿using System;
-
+﻿
 namespace BootCamp.Chapter
 {
     class Program
     {
-        static void Main(string[] args)
+        static void Main()
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5011723)

**Pull request closed without merging**